### PR TITLE
Instructor: add instructor to course: show instructors in alphabetical order #2203

### LINKED ISSUE #2203: Instructor: add instructor to course: show instructors in alphabetical order
~~It is more intuitive if the new instructor is added to the bottom of the list~~

It is better if the instructor list is shown in alphabetical order. Even better if we auto-scroll to the the new instructor after adding a new instructor.

### PR DESCRIPTION
**Pull request by amarlearning**
Fixes #2203

- New comparator method named `compareByName` is added inside `Instructor.java` which is used to sort the Instructor list by name.

- New boolean method named `isInstructorListSortedByName` is added inside `InstructorCourseEditPage.java`. It is used for checking whether the instructor list is sorted or not.

### CODE REVIEW

**Comment by amarlearning** [comment]
Ready for review

**Comment by amarlearning** [comment]
@thenaesh As you can see this test is dependent on `editInstructorIndex` since now the instructor list gets updated alphabetically due to which if any change is done to instructor name, the index of that instructor is changed. 

```java
courseEditPage.verifyInstructorDetails(editInstructorIndex, "New name", "InsCrsEdit.instructor@gmail.tmt",
               false, "", Const.InstructorPermissionRoleNames.INSTRUCTOR_PERMISSION_ROLE_COOWNER);
```
Considering that instrutor list data is constant, we can create & use the `updatedInstructorIndex` and validate this test successfully.

But in future, if we planned to change the test data then this might cause some problem. Let me know what you think

**Comment by amarlearning** [comment]
ready for review

**Comment by amarlearning** [comment]
@thenaesh reminder!

**Comment by thenaesh** [comment]
@amarlearning Removing that part of the test is fine. It's just that there should probably be new tests to ensure that the list is indeed sorted.

**Comment by amarlearning** [comment]
Ready for review

**Comment by amarlearning** [comment]
Ready for review

**Comment by whipermr5** [comment]
@amarlearning What's the output of `browser.driver.findElements(By.xpath("//*[starts-with(@id, 'instructorname')]"));`? Empty?

**Comment by amarlearning** [comment]
@whipermr5 Returns the list of all WebElement whose `Id` starts with `instructorname`. So it actually returns the input element where the instructor's name is saved. 

_[screenshot]_
imageedit_2_5827900856

_The instructor list in the image is not sorted because I am in some other branch :sweat_smile:_

**Comment by whipermr5** [comment]
> Returns the list of all WebElement whose Id starts with instructorname. So it actually returns the input element where the instructor's name is saved.

So where exactly does the test fail when you do it that way?

**Comment by amarlearning** [comment]
I am not able to get the `input['value']` when I tried to this way
```java
List<WebElement> elements = browser.driver.findElements(By.xpath("//*[starts-with(@id, 'instructorname')]"));

for (WebElement element : elements) {
    instructorNames.add(element.getAttribute("value"));
}
```
And then it returns `isSorted = False` and thats why my test fails.

**Comment by whipermr5** [comment]
> So it actually returns the list of input element where the instructor's name is saved.

I know it's supposed to return that (and it does in your browser tool), but is it actually returning that in the Java Selenium code?

**Comment by amarlearning** [comment]
> but is it actually returning that in the Java Selenium code?

Which implementation are you talking about? 

_[this]_
```java
        List<WebElement> elements = browser.driver.findElements(By.xpath("//*[starts-with(@id, 'instructorname')]"));
        for (int i = 1; i < elements.size(); i++) {
            instructorNames.add(browser.driver.findElement(By.id("instructorname" + i)).getAttribute("value"));
        }
```

_[OR this]_
```java
List<WebElement> elements = browser.driver.findElements(By.xpath("//*[starts-with(@id, 'instructorname')]"));

for (WebElement element : elements) {
    instructorNames.add(element.getAttribute("value"));
}
```

**Comment by amarlearning** [comment]
> but is it actually returning that in the Java Selenium code?

@whipermr5 Yes!

**Comment by whipermr5** [comment]
@amarlearning So `element.getAttribute("value")` returns null/empty for the second approach?

**Comment by amarlearning** [comment]
> So `element.getAttribute("value")` returns null/empty for the second approach?

@whipermr5 Just checked and found out that `List<WebElement> elements = browser.driver.findElements(By.xpath("//*[starts-with(@id, 'instructorname')]"));` is returning **9** elements, while we have only **8** instructors present in the page, and the last element in the list was actually the form input [_screenshot_] which was returning null value. That's why the test was failing that way.

screenshot from 2017-07-18 15-03-54

**Comment by whipermr5** [comment]
@amarlearning Yes; good debugging. You have found your own answer now :) Keep this skill up 👍

File: src/main/java/teammates/storage/api/InstructorsDb.java [hunk 1]
```diff
@@ -388,7 +389,9 @@ private Instructor getInstructorEntityForRegistrationKey(String key) {
     }
 
     private List<Instructor> getInstructorEntitiesForCourse(String courseId) {
-        return load().filter("courseId =", courseId).list();
+        List<Instructor> instructorReturnList = load().filter("courseId =", courseId).list();
+        Collections.sort(instructorReturnList, Instructor.compareByName);
+        return instructorReturnList;
```
Comment (whipermr5) [inline src/main/java/teammates/storage/api/InstructorsDb.java]: Please don't do sorting in the storage layer. You can sort the attributes in the logic layer where it's used.

File: src/main/java/teammates/storage/entity/Instructor.java [hunk 2]
```diff
@@ -18,6 +19,15 @@
 public class Instructor extends BaseEntity {
 
     /**
+    * Sort the Instructors list alphabetically by name.
+    */
+    public static Comparator<Instructor> compareByName = new Comparator<Instructor>() {
+        public int compare(Instructor one, Instructor other) {
```
Comment (whipermr5) [inline src/main/java/teammates/storage/entity/Instructor.java]: This should be moved to `InstructorAttributes`, which should implement `Comparable` (use `FeedbackQuestionAttributes` as an example).

**Pull request merged**